### PR TITLE
Support probability, decision function and CSR input format for SVC

### PR DESCRIPTION
**Pull request by RomaA2000**
Support probability, decision function and CSR input format for SVC algorithm

### CODE REVIEW

File: src/daal4py.cpp [hunk 1]
```diff
@@ -427,24 +491,22 @@ daal::data_management::NumericTablePtr make_nt(PyObject * obj)
                 if(PyErr_Occurred()) {PyErr_Print(); throw std::runtime_error("Python Error");}
 
                 if(np_indcs && np_roffs && np_vals && nr && nc) {
-                    // for now, increment indcs by 1
                     size_t * c_indcs = (size_t*)array_data(np_indcs);
                     size_t n = array_size(np_indcs, 0);
-                    for(size_t i=0; i<n; ++i) c_indcs[i] += 1;
-
+                    size_t * c_indcs_one_based = (size_t*)malloc(n * sizeof(size_t));
```
Comment (PetrovKP) [inline src/daal4py.cpp:496]: Use service functions from DAAL: https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/include/services/daal_memory.h

```suggestion
                    size_t * c_indcs_one_based = static_cast<size_t*>(daal::services::daal_malloc(n * sizeof(size_t));
```

File: src/daal4py.cpp [hunk 2]
```diff
@@ -427,24 +491,22 @@ daal::data_management::NumericTablePtr make_nt(PyObject * obj)
...
                     size_t c_nc = (size_t)PyInt_AsSsize_t(nc);
                     if(PyErr_Occurred()) {PyErr_Print(); throw std::runtime_error("Python Error");}
                     size_t c_nr = (size_t)PyInt_AsSsize_t(nr);
                     if(PyErr_Occurred()) {PyErr_Print(); throw std::runtime_error("Python Error");}
-
 #define MKCSR_(_T)                                                      \
                     ret = daal::data_management::CSRNumericTable::create(daal::services::SharedPtr<_T>((_T*)array_data(np_vals), NumpyDeleter((PyArrayObject*)np_vals)), \
-                                                                         daal::services::SharedPtr<size_t>(c_indcs, NumpyDeleter((PyArrayObject*)np_indcs)), \
-                                                                         daal::services::SharedPtr<size_t>(c_roffs, NumpyDeleter((PyArrayObject*)np_roffs)), \
+                                                                         daal::services::SharedPtr<size_t>(c_indcs_one_based, VoidDeleter()), \
```
Comment (PetrovKP) [inline src/daal4py.cpp:508]: Use [service deleter](https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/include/services/daal_shared_ptr.h#L82) from DAAL:
```suggestion
                                                                         daal::services::SharedPtr<size_t>(c_indcs_one_based, daal::services::ServiceDeleter()), \
```

File: src/daal4py.cpp [hunk 3]
```diff
@@ -159,28 +219,33 @@ PyObject * make_nda(daal::data_management::NumericTablePtr * ptr)
...
         if((res = _make_nda_from_homogen<float, NPY_FLOAT32>(ptr)) != NULL) return res;
         if(__oneAPI_imp == 0 && (res = make_py_from_sycltable(ptr, NPY_FLOAT32, (*ptr)->getNumberOfRows(), (*ptr)->getNumberOfColumns())) != Py_None) return res;
-        if((res = _make_nda_from_bd<float, NPY_FLOAT32>(ptr)) != NULL) return res;
+        if((res = _make_nda_from_bd<float, NPY_FLOAT32>(ptr)) != NULL) return res; 
+        if((res = _make_nda_from_csr<float, NPY_FLOAT32>(ptr)) != NULL) return res;
         break;
     case daal::data_management::data_feature_utils::DAAL_INT32_S:
         if((res = _make_nda_from_homogen<int32_t, NPY_INT32>  (ptr)) != NULL) return res;
         if(__oneAPI_imp == 0 && (res = make_py_from_sycltable(ptr, NPY_INT32, (*ptr)->getNumberOfRows(), (*ptr)->getNumberOfColumns())) != Py_None) return res;
-        if((res = _make_nda_from_bd<int32_t, NPY_INT32>(ptr)) != NULL) return res;
```
Comment (PetrovKP) [inline src/daal4py.cpp:171]: Why do you delete this line?

File: src/daal4py.cpp [hunk 4]
```diff
@@ -132,6 +147,51 @@ static PyObject * _make_nda_from_homogen(daal::data_management::NumericTablePtr
...
+static PyObject * _make_nda_from_csr(daal::data_management::NumericTablePtr * ptr)
+{
+    daal::data_management::CSRNumericTable * csr_ptr = dynamic_cast<daal::data_management::CSRNumericTable *>(const_cast<daal::data_management::NumericTable *>((*ptr).get()));
+    if(csr_ptr) {
+        T * data_ptr;
+        size_t * colIndices_ptr;
+        size_t * rowOffsets_ptr;
+        csr_ptr->getArrays<T>(&data_ptr, &colIndices_ptr, &rowOffsets_ptr);
+        size_t n = csr_ptr->getDataSize();
+        T * data_copy = (T*)malloc(n * sizeof(T));
```
Comment (PetrovKP) [inline src/daal4py.cpp:168]: Why doesn't you use  data from `py_data` after allocate ?  Does Numpy have API for allocate data?

File: src/daal4py.cpp [hunk 5]
```diff
@@ -85,6 +99,7 @@ void set_rawp_base(PyArrayObject * ary, void * ptr)
 }
 
 
+
```
Comment (PetrovKP) [inline src/daal4py.cpp:102]: delete extra line

File: src/daal4py.cpp [hunk 6]
```diff
@@ -63,6 +63,20 @@ class NumpyDeleter : public daal::services::DeleterIface
     PyArrayObject* _ndarray;
 };
 
+class VoidDeleter : public daal::services::DeleterIface
```
Comment (PetrovKP) [inline src/daal4py.cpp:66]: He's not needed

File: generator/wrappers.py [hunk 7]
```diff
@@ -519,7 +525,6 @@ def wrap_algo(algo, ver):
     'algorithms::math::smoothrelu': ['ParameterType',],
     'algorithms::math::softmax': ['ParameterType',],
     'algorithms::math::tanh': ['ParameterType',],
-    'algorithms::multi_class_classifier': ['Result',],
```
Comment (PetrovKP) [inline generator/wrappers.py:522]: Please, revert

File: generator/wrappers.py [hunk 8]
```diff
@@ -205,6 +206,11 @@ def wrap_algo(algo, ver):
     'algorithms::gbt::classification::training::varImportance': 'algorithms::gbt::training::VariableImportanceModes',
     'algorithms::gbt::regression::training::varImportance': 'algorithms::gbt::training::VariableImportanceModes',
 }
+result_to_compue = {
+    'algorithms::multi_class_classifier::prediction' : 'algorithms::multi_class_classifier::ResultToComputeId',
+}
+
+
```
Comment (PetrovKP) [inline generator/wrappers.py:213]: delete extra line

File: generator/wrappers.py [hunk 9]
```diff
@@ -139,7 +139,8 @@ def wrap_algo(algo, ver):
     'algorithms::kdtree_knn_classification': ['seed',], # deprecated
     'algorithms::lasso_regression::training': ['optionalArgument'], # internal stuff
     'algorithms::lasso_regression::prediction': ['algorithms::linear_model::interceptFlag',], # parameter
-    'algorithms::multi_class_classifier': ['algorithms::multi_class_classifier::getTwoClassClassifierModels',], # unsupported return type ModelPtr*
+    'algorithms::multi_class_classifier': ['algorithms::multi_class_classifier::getTwoClassClassifierModels', 'algorithms::classifier::ResultToComputeId'], # unsupported return type ModelPtr*
```
Comment (PetrovKP) [inline generator/wrappers.py:142]:
```suggestion
    'algorithms::multi_class_classifier': ['algorithms::multi_class_classifier::getTwoClassClassifierModels'], # unsupported return type ModelPtr*
```

File: generator/wrappers.py [hunk 10]
```diff
@@ -139,7 +139,8 @@ def wrap_algo(algo, ver):
...
+    'algorithms::multi_class_classifier::prediction': ['algorithms::classifier::prediction::ResultId', 'algorithms::classifier::prediction::Result', 'algorithms::classifier::ResultToComputeId'],
```
Comment (PetrovKP) [inline generator/wrappers.py:143]:
```suggestion
    'algorithms::multi_class_classifier::prediction': ['algorithms::classifier::prediction::ResultId', 'algorithms::classifier::prediction::Result'],
```

File: generator/gen_daal4py.py [hunk 11]
```diff
@@ -644,6 +647,7 @@ def prepare_hlwrapper(self, ns, mode, func, no_dist, no_stream):
             param_classes = self.get_all_parameter_classes(ns)
             all_params = OrderedDict()
             opt_params = {}
+
```
Comment (PetrovKP) [inline generator/gen_daal4py.py:650]:
```suggestion
```

File: generator/gen_daal4py.py [hunk 12]
```diff
@@ -147,7 +147,8 @@ def read(self):
                     #print('reading ' +  fname)
                     with open(fname, "r") as header:
                         parsed_data = parse_header(header, cython_interface.ignores)
-
+                    # if parsed_data['enums']:
```
Comment (PetrovKP) [inline generator/gen_daal4py.py:150]: delete commetes

File: daal4py/sklearn/cluster/k_means.py [hunk 13]
```diff
@@ -15,12 +15,12 @@
...
+# from sklearn import __version__ as sklearn_version
+# from distutils.version import LooseVersion
 
-if LooseVersion(sklearn_version) >= LooseVersion("0.23"):
-    from ._k_means_0_23 import *
-elif LooseVersion(sklearn_version) >= LooseVersion("0.22"):
-    from ._k_means_0_22 import *
-else:
-    from ._k_means_0_21 import *
+# if LooseVersion(sklearn_version) >= LooseVersion("0.23"):
```
Comment (PetrovKP) [inline daal4py/sklearn/cluster/k_means.py:21]: revert changes

**Review by PetrovKP — DISMISSED** [review]

File: src/daal4py.cpp [hunk 14]
```diff
@@ -323,7 +373,7 @@ daal::data_management::NumericTablePtr make_nt(PyObject * obj)
 #define MAKENT_(_T) ptr = _make_hnt<_T>(obj)
                 SET_NPY_FEATURE(PyArray_DESCR(ary)->type, MAKENT_, throw std::invalid_argument("Found unsupported array type"));
 #undef MAKENT_
-            } else {
+            } else {
```
Comment (owerbat) [inline src/daal4py.cpp:376]: Unnecessary tab

File: daal4py/sklearn/svm/_svm_0_23.py [hunk 15]
```diff
@@ -20,23 +20,31 @@
...
 import daal4py
-from .._utils import (make2d, getFPType, method_uses_sklearn, method_uses_daal, daal_check_version)
+from .._utils import (make2d, getFPType, method_uses_sklearn,
+                      method_uses_daal, daal_check_version)
 import logging
 
 
 LIBSVM_IMPL = ['c_svc', 'nu_svc', 'one_class', 'epsilon_svr', 'nu_svr']
 
+
```
Comment (owerbat) [inline daal4py/sklearn/svm/_svm_0_23.py:47]: Do we want to make all these changes for 0.21 and 0.22 versions?

**Review by owerbat — APPROVED** [review]

**Pull request merged**